all/experiments/parallel_env_experiment.py

```diff
@@ -56,88 +63,59 @@ def episode(self):
+                start_time = cur_time
+                for i in range(num_envs):
+                    if dones[i]:
+                        self._log_training_episode(returns[i], fps)
+                        returns[i] = 0
+            self._episode += episodes_completed
 
     def test(self, episodes=100):
         test_agent = self._preset.test_agent()
-        env = self._envs[0].duplicate(1)[0]
-        returns = []
-        for episode in range(episodes):
-            episode_return = self._run_test_episode(test_agent, env)
-            returns.append(episode_return)
-            self._log_test_episode(episode, episode_return)
-        self._log_test(returns)
-        return returns
-
-    def _reset(self):
-        for env in self._envs:
-            env.reset()
-        rewards = torch.zeros(
-            (self._n_envs),
-            dtype=torch.float,
-            device=self._envs[0].device
-        )
-        self._returns = rewards
-        self._episode_start_times = [timer()] * self._n_envs
-        self._episode_start_frames = [self._frame] * self._n_envs
-
-    def _step(self):
-        states = self._aggregate_states()
-        actions = self._agent.act(states)
-        self._step_envs(actions)
-
-    def _step_envs(self, actions):
-        for i, env in enumerate(self._envs):
-            state = env.state
-            if self._render:
-                env.render()
-
-            if state.done:
-                self._returns[i] += state.reward
-                self._log_training_episode(self._returns[i].item(), self._fps(i))
-                env.reset()
-                self._returns[i] = 0
-                self._episode += 1
-                self._episode_start_times[i] = timer()
-                self._episode_start_frames[i] = self._frame
-            else:
-                action = actions[i]
-                if action is not None:
-                    self._returns[i] += state.reward
-                    env.step(action)
-                    self._frame += 1
-
-    def _aggregate_states(self):
-        return State.array([env.state for env in self._envs])
-
-    def _aggregate_rewards(self):
-        return torch.tensor(
-            [env.state.reward for env in self._envs],
-            dtype=torch.float,
-            device=self._envs[0].device
-        )
-
-    def _run_test_episode(self, test_agent, env):
-        # initialize the episode
-        state = env.reset()
-        action = test_agent.act(state)
         returns = 0
-
-        # loop until the episode is finished
-        while not state.done:
-            if self._render:
-                env.render()
-            state = env.step(action)
-            action = test_agent.act(state)
-            returns += state.reward
-
-        return returns
-
-    def _fps(self, i):
-        end_time = timer()
-        return (self._frame - self._episode_start_frames[i]) / (end_time - self._episode_start_times[i])
+        first_state = self._env.reset()[0]
```

Comment:
Will definitely have to come back to this soon, as discussed.

Reply:
Yes, this is really hacky.